ehrql/renderers.py

```diff
@@ -1,134 +1,75 @@
+        return sequence, False, []
 
     # Check we have enough rows to truncate to head and tail
-    if rows.count("<tr>") <= ((head or 0) + (tail or 0)):
-        return table_repr
+    if len(sequence) <= head + tail:
+        return sequence, False, []
 
-    # split row on <tr> tokens, remove any empty strings (this will lose the first <tr> of
-    # each row, but we'll add it in again later)
-    rows = [row for row in rows.split("<tr>") if row]
-    # compose an "ellipsis row" to mark the place of truncated rows
-    td_count = rows[0].count("<td>")
-    ellipsis_row = f"{'<td>...</td>' * td_count}</tr>"
+    head_rows = []
+    tail_rows = []
 
-    # Build the list of rows we need to include, with ellipsis rows where necessary
-    truncated_rows = []
+    if head:
+        head_rows = sequence[:head]
 
-    head_rows = rows[:head] if head is not None else [ellipsis_row]
-    truncated_rows.extend(head_rows)
+    if tail:
+        tail_rows = sequence[-tail:]
 
-    if head is not None and tail is not None:
-        truncated_rows.append(ellipsis_row)
+    return head_rows, True, tail_rows
 
-    tail_rows = rows[-tail:] if tail is not None else [ellipsis_row]
-    truncated_rows.extend(tail_rows)
 
-    # re-join the truncated rows with <tr>
-    truncated_rows = "<tr>" + "<tr>".join(truncated_rows)
-    return start + truncated_rows + end
+def records_to_html_table(records: list[dict], head: int = 0, tail: int = 0):
+    rows = []
+    headers = "".join([f"<th>{header}</th>" for header in records[0].keys()])
+    head_rows, ellipsis, tail_rows = headtail(records, head, tail)
 
+    def html_row(row):
+        columns = "".join(f"<td>{v}</td>" for v in row.values())
+        return f"<tr>{columns}</tr>"
 
-def _truncate_lines(
-    table_repr: str, headers: int = 0, head: int | None = None, tail: int | None = None
-):
-    table_rows = [row for row in table_repr.split("\n") if row]
+    for row in head_rows:
+        rows.append(html_row(row))
 
-    # Check we have enough rows to truncate to head and tail
-    if len(table_rows) <= (headers + (head or 0) + (tail or 0)):
-        return table_repr
+    if ellipsis:
+        ellipsis_columns = "<td>&hellip;</td>" * len(records[0])
+        rows.append(f"<tr>{ellipsis_columns}</tr>")
 
-    # compose an "ellipsis row" to mark the place of truncated rows
-    cell_count = table_rows[0].count("|") + 1
-    ellipsis_row = " | ".join("...".ljust(17) for i in range(cell_count)).strip()
+    for row in tail_rows:
+        rows.append(html_row(row))
 
-    # Build the list of rows we need to include, with ellipsis rows where necessary
-    truncated_rows = table_rows[:headers]
+    html_rows = "".join(rows)
 
-    head_rows = (
-        table_rows[headers : headers + head] if head is not None else [ellipsis_row]
-    )
-    truncated_rows.extend(head_rows)
+    return f"{START_MARKER}<table><thead><tr>{headers}</tr></thead><tbody>{html_rows}</tbody></table>{END_MARKER}"
```

Comment:
Do we still need the `START_MARKER` and `END_MARKER`? Or were they just so the previous regex would find the table?

Reply:
Yes, they are not for the regex, but for the vscode extension to know where pre-formatted html tables start and end (this bit is what I am working on now).

Reply:
Ah of course. I'd naively just searched the ehrql code space for the marker and forgot about the extension.